File: 138 - Copy List with Random Pointer/hasmap.go
func copyRandomList(head *Node) *Node {
	current := head
	values := make(map[*Node]*Node)

	for current != nil {
		n := &Node{current.Val, nil, nil}
		values[current] = n
		current = current.Next
	}

	current = head
	for current != nil {
		newNode := values[current]
		newNode.Random = values[current.Random]
		newNode.Next = values[current.Next]
		current = current.Next
	}

	return values[head]
}

File: 138 - Copy List with Random Pointer/hasmap.java
class Solution {
    public Node copyRandomList(Node head) {
        
        if(head == null){ 
            return head;
        }

        Node current = head;
        Map<Node, Node> nodes = new HashMap<Node, Node>();

        while(current != null) {
            Node node = new Node(current.val, null, null);
            nodes.put(current, node);
            current = current.next;
        }

        nodes.forEach((key, value) -> {
            value.next = nodes.get(key.next);
            value.random = nodes.get(key.random);
        });

        return nodes.get(head);
    }
}